### ISSUE TITLE
#887 Recorder changes output format for complex numbers,


### ISSUE THREAD

**Issue by nikhilgupta10** [gridlab-d/gridlab-d#887]
The recorder will sometimes change format for complex numbers, particularly 'i' vs. 'j' after the first time step.  This may indicate that it grabs the initial value in the objects format, then switches to the recorder's imposed format at a later period.  This may be minor, but may by indicative of further issues.

See:  https://sourceforge.net/p/gridlab-d/discussion/842562/thread/84c18d4b/?limit=25,

**Comment by dpinney** [comment]
Sometimes 'd' is also used to represent sqrt(-1). I've seen it happen for triplex_meter voltage_12 values.

**Comment by dchassin** [comment]
It should be noted that some of the complex math functions "touch" the notation and can change a property from rectangular to polar or vice-versa.

**Comment by dchassin** [comment]
I wouldn't be surprised if for small angles it is difficult to tell the difference between 'd' and 'i', or 'j' because sin(y) is nearly equal to y. In such cases it would certainly look like 'd' represents sqrt(-1).

**Comment by dpinney** [comment]
I didn't realize GLD was using x+yd to denote polar form. Strange that the default for most values is rectangular, but polar is used for triplex meters. Would be nice to have the same default everywhere.

**Comment by dchassin** [comment]
I think it uses whatever the last operation that touched thinks is appropriate. I've always wanted to have direct control over both the unit and the formatting the way the server does. I believe that will have to wait until the tape module gets upgraded to use the new API.

D.

On Oct 19, 2017, at 17:51, David Pinney <notifications@github.com<mailto:notifications@github.com>> wrote:


I didn't realize GLD was using x+yd to denote polar form. Strange that the default for most values is rectangular, but polar is used for triplex meters. Would be nice to have the same default everywhere.

**Comment by dchassin** [comment]
See issue #1103 for additional information on polar/rect notation problems in tape/recorder.c

**Comment by ftuffner** [comment]
Added in a global to set this in the "convert_from_complex" function, which I believe everything uses.  Seems to work fine.  The global is `complex_output_format`, with possible options of:

- `DEFAULT` – same as existing behavior (dictated by the variable)
- `RECT` – complex numbers should be in rectangular format
- `POLAR_DEG` – complex numbers should be in polar format, with the angles in degrees
- `POLAR_RAD` – complex numbers should be in polar format, with the angles in radians

So example usage would be: `#set complex_output_format=RECT`

Works fine under Windows and Linux (Ubuntu Focal) so far.  Planning to test Mac later tonight/tomorrow to make sure it works, then will file a PR.

**Comment by ftuffner** [comment]
Confirmed works on MacOS as well.  Waiting on one final test, then will submit this as PR into develop/4.3
